File: services/memory_agent.ts
import { getContext } from "./context_engine";
import { saveMemory } from "./vector_memory";
import { logTask, formatNote, saveNote } from "./hybrid_memory";

export interface TaskResult {
  success: boolean;
  result: string;
  context: string;
  metadata: {
    taskId: string;
    timestamp: string;
    duration: number;
  };
}

export class MemoryAgent {
  private agentName: string;
  private autoSave: boolean = true;

  constructor(agentName: string) {
    this.agentName = agentName;
  }

  async execute(
    task: string,
    executorFn: (taskWithContext: string) => Promise<string>
  ): Promise<TaskResult> {
    const startTime = Date.now();
    const taskId = `${this.agentName}_${Date.now()}`;

    const contextResult = await getContext(task);
    const taskWithContext = this.buildContextualTask(task, contextResult.context);

    let result: string;
    let success = true;

    try {
      result = await executorFn(taskWithContext);
    } catch (error) {
      result = `Erro: ${error instanceof Error ? error.message : "Erro desconhecido"}`;
      success = false;
    }

    const duration = Date.now() - startTime;

    if (this.autoSave) {
      await this.saveExperience(taskId, task, result, taskWithContext, success);
    }

    return {
      success,
      result,
      context: contextResult.context,
      metadata: {
        taskId,
        timestamp: new Date().toISOString(),
        duration,
      },
    };
  }

  private buildContextualTask(task: string, context: string): string {
    if (!context || context.includes("Nenhuma memória")) {
      return task;
    }

    return `Contexto anterior:\n${context}\n---\nNova tarefa:\n${task}`;
  }

  private async saveExperience(
    taskId: string,
    originalTask: string,
    result: string,
    contextUsed: string,
    success: boolean
  ): Promise<void> {
    const fullContent = formatNote(originalTask, result);
    const logEntry = `${fullContent}\n\n---\n\n### Contexto utilizado:\n${contextUsed}\n\n### Sucesso: ${success}`;

    await saveMemory(taskId, `${originalTask}\n${result}\n${contextUsed}`);

    await saveNote(
      "05-Logs",
      taskId,
      logEntry
    );
  }

  setAutoSave(enabled: boolean): void {
    this.autoSave = enabled;
  }

  async executeSimple(
    task: string,
    executorFn: (taskWithContext: string) => Promise<string>
  ): Promise<TaskResult> {
    return this.execute(task, executorFn);
  }
}

export async function runWithMemory(
  agentName: string,
  task: string,
  executorFn: (taskWithContext: string) => Promise<string>
): Promise<TaskResult> {
  const agent = new MemoryAgent(agentName);
  return agent.execute(task, executorFn);
}

export async function quickLog(task: string, result: string): Promise<void> {
  await logTask(task, result);
}

File: antigravity-backend/services/memory_agent.py
from datetime import datetime

from services.context_engine import get_context
from services.vector_memory import save_memory
from services.hybrid_memory import log_task


class MemoryAgent:
    def __init__(self, agent_name: str):
        self.agent_name = agent_name
        self.auto_save = True

    async def execute(self, task: str, executor_fn):
        task_id = f"{self.agent_name}_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
        context_result = await get_context(task)
        
        task_with_context = self._build_contextual_task(task, context_result["context"])
        
        result = await executor_fn(task_with_context)
        
        if self.auto_save:
            await self._save_experience(task_id, task, result, context_result["context"])
        
        return {
            "success": True,
            "result": result,
            "context": context_result["context"],
            "metadata": {
                "task_id": task_id,
                "timestamp": datetime.now().isoformat()
            }
        }

    def _build_contextual_task(self, task: str, context: str) -> str:
        if not context or "Nenhuma memória" in context:
            return task
        return f"Contexto anterior:\n{context}\n---\nNova tarefa:\n{task}"

    async def _save_experience(self, task_id: str, task: str, result: str, context_used: str):
        full_content = f"{task}\n{result}\n\nContexto: {context_used}"
        await save_memory(task_id, full_content)
        await log_task(task_id, result)


async def run_with_memory(agent_name: str, task: str, executor_fn):
    agent = MemoryAgent(agent_name)
    return await agent.execute(task, executor_fn)